### DISCUSSION TITLE
RIC action definition format 4

### DISCUSSION THREAD

**Discussion by yc541** [srsran/srsRAN_Project#438] [General Help]
There may be some mismatch between the asn1 encoder/decoder implemented in srsRAN and FlexRIC. I did a bit more digging following https://github.com/srsran/srsRAN_Project/discussions/437
In the function _e2_sm_kpm_action_definition_s::unpack(cbit_ref& bref)_, _ric_style_type_ is correctly decoded to 4, but in the next line in function _e2_sm_kpm_action_definition_s::action_definition_formats_c_::unpack(cbit_ref& bref)_, the line _e.unpack(bref)_ decodes it to 2 rather than 4. 
To get around with this I slightly modified _e2_sm_kpm_action_definition_s::action_definition_formats_c_::set(types::options e)_ to get it to switch based on _ric_style_type_ rather than _type__ (the values of _type__ are correctly assigned as well within the cases). With the changes the code initializes the correct action definition format 4 but there are still subsequent issues unpacking the AD (not sure if the changes are the cause). In the line _HANDLE_CODE(unpack_dyn_seq_of(matching_ue_cond_list, bref, 1, 32768, true))_, the _nof_items_(in the function definition in asn1_util.h) is decoded to 13569, which should probable be 1 since in the xapp's code only _dst.matching_cond_lst[0]_ is defined. 

I have attached the pcap of AD format 1 and 4.
[pcap.zip](https://github.com/srsran/srsRAN_Project/files/14112764/pcap.zip)

**Comment by HaoxinSEU** [comment]
Hi @yc541,

I'm also trying to make KPM REPORT style 4 work in srsRAN with the `xapp_oran_moni` in FlexRIC. But I think I encounter the same problem as you. Do you make any progress about this decoding? Or do you have more hints about how to go further?

Thanks!

**Reply by yc541** [reply]
Hello, I have not done any further debugging of this issue.

**Comment by pgawlowicz** [comment]
Hi @yc541,

We have just released a repository that allows running ORAN-SC-RIC as a multi-container Docker application with a single docker command. You can access the repository at: https://github.com/srsran/oran-sc-ric

Moreover, the repository includes two Python-based xApps examples designed to facilitate the requesting and monitoring of E2 metrics from the gNB. These examples cover all reporting styles (1-5) of the e2sm-kpm module, which has been updated to version 3.

To get started quickly, refer to the `Quick Start` section in the readme file.
Please note that until the next code release, you'll need to utilize the `test` branch from the srsRAN_Project repository.

We are currently working on an example xApp using e2sm_rc module.
